### PR TITLE
Full Video Labels endpoint

### LINKED ISSUE #452: getVideoLabel endpoint

### PR DESCRIPTION
**Pull request by mchangrh**
closes #452 

- [x] by query param
- [x] by hash
- [x] [Draft Documentation](https://wiki.sponsor.ajay.app/index.php?title=API_Docs/Draft&oldid=947)

### CODE REVIEW

File: src/routes/getVideoLabel.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        return chooseSegment(segments);
+    } catch (err) {
+        if (err) {
+            Logger.error(err as string);
+            return null;
+        }
+    }
+}
+
+async function getLabelsbyHash(hashedVideoIDPrefix: VideoIDHash, service: Service): Promise<SBRecord<VideoID, VideoData>> {
```
Comment (ajayyy) [inline src/routes/getVideoLabel.ts:37]:
```suggestion
async function getLabelsByHash(hashedVideoIDPrefix: VideoIDHash, service: Service): Promise<SBRecord<VideoID, VideoData>> {
```

File: src/routes/getVideoLabel.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        choices = locked;
+    }
+    //no need to filter, just one label
+    if (choices.length === 1) {
+        return transformDBSegments(choices);
+    }
+    // sponsor > exclusive > selfpromo
+    const sponsorResult = choices.find((segment) => segment.category === "sponsor");
+    const eaResult = choices.find((segment) => segment.category === "exclusive_access");
+    const selfpromoResult = choices.find((segment) => segment.category === "selfpromo");
```
Comment (ajayyy) [inline src/routes/getVideoLabel.ts:124]: maybe convert these to lambda functions to only run as many as are needed

Comment (mchangrh) [reply src/routes/getVideoLabel.ts]:
```ts
 const categoryResult = findCategory("sponsor") ?? findCategory("exclusive_access") ?? findCategory("selfpromo");
    if (categoryResult) results.push(categoryResult);
```

**Comment by ajayyy** [comment]
Looks like redis is never reset.

File: src/utils/queryCacher.ts [hunk 3]
```diff
@@ -81,6 +81,7 @@ function clearSegmentCache(videoInfo: { videoID: VideoID; hashedVideoID: VideoID
         redis.del(skipSegmentsKey(videoInfo.videoID, videoInfo.service)).catch((err) => Logger.error(err));
         redis.del(skipSegmentGroupsKey(videoInfo.videoID, videoInfo.service)).catch((err) => Logger.error(err));
         redis.del(skipSegmentsHashKey(videoInfo.hashedVideoID, videoInfo.service)).catch((err) => Logger.error(err));
+        redis.del(videoLabelsKey(videoInfo.hashedVideoID, videoInfo.service)).catch((err) => Logger.error(err));
```
Comment (ajayyy) [inline src/utils/queryCacher.ts:84]: videoLabelsHashKey

File: src/utils/queryCacher.ts [hunk 4]
```diff
@@ -81,6 +81,7 @@ function clearSegmentCache(videoInfo: { videoID: VideoID; hashedVideoID: VideoID
         redis.del(skipSegmentsKey(videoInfo.videoID, videoInfo.service)).catch((err) => Logger.error(err));
         redis.del(skipSegmentGroupsKey(videoInfo.videoID, videoInfo.service)).catch((err) => Logger.error(err));
         redis.del(skipSegmentsHashKey(videoInfo.hashedVideoID, videoInfo.service)).catch((err) => Logger.error(err));
+        redis.del(videoLabelsHashKey(videoInfo.hashedVideoID, videoInfo.service)).catch((err) => Logger.error(err));
```
Comment (ajayyy) [inline src/utils/queryCacher.ts:84]: you need both

Comment (mchangrh) [reply src/utils/queryCacher.ts:84]: oh right

**Pull request merged**